### PR TITLE
Initial work on deep linking and web support

### PR DESCRIPTION
**Pull request by satya164**
Please rebase and merge instead of squash when merging :)

I have made rehydration more robust to add all missing fields and handle invalid state more gracefully as we discussed before. This means you can pass as minimum of data as needed as initial state.

For example:

```js
{
  routes: [
    {
      name: 'foo',
      state: {
        routes: [
          {
            name: 'bar',
            params: { count: 42 },
          },
        ],
      },
    },
  ],
}
```

I have also added 2 helper utilities for converting between URL and state. Converting is lossy because it's not really possible to keep all data in the URL (e.g. params for non-terminal route). But this kind of data can be stored in the history object with `history.pushState` etc.

I'm not very happy with the utilities currently because of the way it handles params. Currently it'll `JSON.stringify` the params which works, but looks ugly in the URL. However we can think of how to improve it later.

### CODE REVIEW

File: packages/core/src/getStateFromPath.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    } else {
+      result = state;
+    }
+
+    current = state;
+    segments.shift();
+  }
+
+  if (current == null || result == null) {
+    throw new Error(`Unable to parse path to state: ${path}`);
```
Comment (osdnk) [inline packages/core/src/getStateFromPath.tsx]: I'm not sure if we want to throw an exception here. It's not problem of our code, that given url is broken. Maybe we should fallback to initial state?

File: packages/core/src/getStateFromPath.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ */
+export default function getStateFromPath(path: string): InitialState {
+  const parts = path.split('?');
+  const segments = parts[0].split('/').filter(Boolean);
+  const query = parts[1] ? parts[1].split('&') : undefined;
+
+  let result: InitialState | undefined;
+  let current: InitialState | undefined;
+
+  while (segments.length) {
```
Comment (osdnk) [inline packages/core/src/getStateFromPath.tsx:19]: Can't we make it with reduce?

File: packages/routers/src/StackRouter.tsx [hunk 3]
```diff
@@ -47,36 +48,69 @@ export default function StackRouter(options: StackRouterOptions) {
...
         key: `stack-${shortid()}`,
-        index,
+        index: 0,
         routeNames,
-        routes: routeNames.slice(0, index + 1).map(name => ({
-          name,
-          key: `${name}-${shortid()}`,
-          params: routeParamList[name],
-        })),
+        routes: [
```
Comment (osdnk) [inline packages/routers/src/StackRouter.tsx:60]: makes sense, but AFAIK it's unrelated? Can you confirm?

File: packages/core/src/getPathFromState.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { NavigationState, PartialState, Route } from './types';
+
+type State = NavigationState | Omit<PartialState<NavigationState>, 'stale'>;
+
+/**
+ * Utility to serialize a navigation state object to a path string.
+ * This is a lossy operation because we can't store all of the data in the path.
```
Comment (osdnk) [inline packages/core/src/getPathFromState.tsx]: This comment maybe is not needed? `Utility to serialize a navigation state object to a path string.` should be fine

File: packages/core/src/__tests__/getStateFromPath.test.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+              params: { count: 42 },
+            },
+          ],
+        },
+      },
+    ],
+  });
+});
+
+it('handles ending slash when converting', () => {
```
Comment (osdnk) [inline packages/core/src/__tests__/getStateFromPath.test.tsx:55]: 😍

File: packages/core/src/getStateFromPath.tsx (hunk 2 above)
Comment (satya164) [reply packages/core/src/getStateFromPath.tsx:19]: We can, but I feel it's simpler with a loop.

File: packages/routers/src/StackRouter.tsx (hunk 3 above)
Comment (satya164) [reply packages/routers/src/StackRouter.tsx:60]: yeah, not strictly related

**Pull request merged**